### ISSUE TITLE
Unable to create a member (against existing/migrated v8 database)

### ISSUE THREAD

**Issue by adamhearn** [umbraco/Umbraco-CMS#10666]
### Which *exact* Umbraco version are you using? For example: 8.13.1 - don't just write v8

9.0.0-rc001

### Bug summary

Existing v8 site being migrated to v9. It seems that my existing logic of adding a new member is not working in the RC of v9.

The member UI works (can create new members) as expected.

I get no errors/exceptions in the view where the code is being invoked from.

### Specifics

In my view I have:
```
@inject IMemberService _memberService

_memberService.CreateMemberWithIdentity( User.Identity.Name, String.Format( "{0}@domain.com", empId ), System.Environment.UserName, "Member" );
```
**empid** is just a string name calculated prior to the call

It goes through Umbraco layers without issue and no error is raised. However, a SQL trace shows that the following SQL statements are executed:

```
exec sp_executesql N'UPDATE umbracoLock WITH (REPEATABLEREAD) SET value = (CASE WHEN (value=1) THEN -1 ELSE 1 END) WHERE id=@0',N'@0 int',@0=-335
go
exec sp_executesql N'SELECT value FROM umbracoLock WITH (REPEATABLEREAD) WHERE id=@0',N'@0 int',@0=-337
go
exec sp_executesql N'SELECT [umbracoNode].[id] AS [NodeId], [umbracoNode].[uniqueId] AS [UniqueId], [umbracoNode].[parentId] AS [ParentId], [umbracoNode].[level] AS [Level], [umbracoNode].[path] AS [Path], [umbracoNode].[sortOrder] AS [SortOrder], [umbracoNode].[trashed] AS [Trashed], [umbracoNode].[nodeUser] AS [UserId], [umbracoNode].[text] AS [Text], [umbracoNode].[nodeObjectType] AS [NodeObjectType], [umbracoNode].[createDate] AS [CreateDate]
FROM [umbracoNode]
WHERE (([umbracoNode].[id] = @0))',N'@0 int',@0=-1
go
exec sp_executesql N'SELECT MAX(sortOrder)
FROM [umbracoNode]
WHERE ((([umbracoNode].[nodeObjectType] = @0) AND ([umbracoNode].[parentId] = @1)))',N'@0 uniqueidentifier,@1 int',@0='39EB0F98-B348-42A1-8662-E7EB18487560',@1=-1
go
exec sp_executesql N'SELECT [umbracoNode].[id] AS [NodeId]
FROM [umbracoNode]
WHERE ((([umbracoNode].[uniqueId] = @0) AND ([umbracoNode].[nodeObjectType] = @1)))',N'@0 uniqueidentifier,@1 uniqueidentifier',@0='ADC0E61B-7BA0-4EF3-ABA3-9930E8E6AB00',@1='92849B1E-3904-4713-9356-F646F87C25F4'
go
exec sp_executesql N'INSERT INTO [umbracoNode] ([uniqueId],[parentId],[level],[path],[sortOrder],[trashed],[nodeUser],[text],[nodeObjectType],[createDate]) VALUES (@0,@1,@2,@3,@4,@5,@6,@7,@8,@9);SELECT SCOPE_IDENTITY();',N'@0 uniqueidentifier,@1 int,@2 smallint,@3 nvarchar(4000),@4 int,@5 int,@6 int,@7 nvarchar(4000),@8 uniqueidentifier,@9 datetime',@0='ADC0E61B-7BA0-4EF3-ABA3-9930E8E6AB00',@1=-1,@2=1,@3=N'-1',@4=25,@5=0,@6=-1,@7=N'adam',@8='39EB0F98-B348-42A1-8662-E7EB18487560',@9='2021-07-13 17:03:50.877'
go
exec sp_executesql N'UPDATE [umbracoNode] SET [uniqueId] = @0, [parentId] = @1, [level] = @2, [path] = @3, [sortOrder] = @4, [trashed] = @5, [nodeUser] = @6, [text] = @7, [nodeObjectType] = @8, [createDate] = @9 WHERE [id] = @10',N'@0 uniqueidentifier,@1 int,@2 smallint,@3 nvarchar(4000),@4 int,@5 int,@6 int,@7 nvarchar(4000),@8 uniqueidentifier,@9 datetime,@10 int',@0='ADC0E61B-7BA0-4EF3-ABA3-9930E8E6AB00',@1=-1,@2=1,@3=N'-1,1149',@4=25,@5=0,@6=-1,@7=N'adam',@8='39EB0F98-B348-42A1-8662-E7EB18487560',@9='2021-07-13 17:03:50.877',@10=1149
go
exec sp_executesql N'INSERT INTO [umbracoContent] ([nodeId],[contentTypeId]) VALUES (@0,@1)',N'@0 int,@1 int',@0=1149,@1=1044
go
exec sp_executesql N'INSERT INTO [umbracoContentVersion] ([nodeId],[versionDate],[userId],[current],[text]) VALUES (@0,@1,@2,@3,@4);SELECT SCOPE_IDENTITY();',N'@0 int,@1 datetime,@2 nvarchar(4000),@3 int,@4 nvarchar(4000)',@0=1149,@1='2021-07-13 17:03:50.877',@2=NULL,@3=1,@4=N'adam'
go
exec sp_executesql N'INSERT INTO [cmsMember] ([nodeId],[Email],[LoginName],[Password],[passwordConfig],[securityStampToken],[emailConfirmedDate]) VALUES (@0,@1,@2,@3,@4,@5,@6)',N'@0 int,@1 nvarchar(4000),@2 nvarchar(4000),@3 nvarchar(4000),@4 nvarchar(4000),@5 nvarchar(4000),@6 nvarchar(4000)',@0=1149,@1=N'adam@domain.com',@2=N'DOMAIN\adam',@3=N'___UIDEMPTYPWORD__AQAAAAEAACcQAAAAEGYBsNLxBaaWeMnAiCtmr2MwIfECQMmqu+AytxZDvE86yRjj73GuWpTH/Q0m22Bo5g==',@4=N'{"hashAlgorithm":"PBKDF2.ASPNETCORE.V3"}',@5=N'0e65f802-621e-4b9a-9a44-ecfd745671b0',@6=NULL
go
exec sp_executesql N'INSERT INTO [umbracoPropertyData] ([versionId],[propertyTypeId],[languageId],[segment],[intValue],[decimalValue],[dateValue],[varcharValue],[textValue]) VALUES (@0,@1,@2,@3,@4,@5,@6,@7,@8);SELECT SCOPE_IDENTITY();',N'@0 int,@1 int,@2 nvarchar(4000),@3 nvarchar(4000),@4 int,@5 nvarchar(4000),@6 nvarchar(4000),@7 nvarchar(4000),@8 nvarchar(4000)',@0=189,@1=30,@2=NULL,@3=NULL,@4=1,@5=NULL,@6=NULL,@7=NULL,@8=NULL
go
exec sp_executesql N'DELETE [umbracoRelation]
FROM [umbracoRelation]
INNER JOIN [umbracoRelationType]
ON [umbracoRelation].[relType] = [umbracoRelationType].[id]
WHERE (([umbracoRelation].[parentId] = @0))
AND ([umbracoRelationType].[alias] IN (@1,@2))',N'@0 int,@1 nvarchar(4000),@2 nvarchar(4000)',@0=1149,@1=N'umbMedia',@2=N'umbDocument'
go
exec sp_executesql N'UPDATE [cmsContentNu] SET data=@0, dataRaw=@1, rv=rv+1 WHERE nodeId=@2 AND published=@3',N'@0 nvarchar(4000),@1 varbinary(195),@2 int,@3 int',@0=NULL,@1=0x92C70362CD0128C6000000B7F90C9389B5756D627261636F4D656D626572436F6D6D656E747390D9231800FA084661696C656450617373776F7264417474656D707473903D00F903417070726F7665649193A0A0D200000001B64400BA4C6F636B65644F757490BC1800306173741C008B6F757444617465903600011E003C67696E9300444C6173749100614368616E67653E0029D9265D00042200FF0552657472696576616C5175657374696F6E90D92429000BD0416E737765729080A46164616D,@2=1149,@3=0
go
exec sp_executesql N'INSERT INTO [cmsContentNu] ([nodeId],[published],[data],[rv],[dataRaw]) VALUES (@0,@1,@2,@3,@4)',N'@0 int,@1 int,@2 nvarchar(4000),@3 bigint,@4 varbinary(195)',@0=1149,@1=0,@2=NULL,@3=0,@4=0x92C70362CD0128C6000000B7F90C9389B5756D627261636F4D656D626572436F6D6D656E747390D9231800FA084661696C656450617373776F7264417474656D707473903D00F903417070726F7665649193A0A0D200000001B64400BA4C6F636B65644F757490BC1800306173741C008B6F757444617465903600011E003C67696E9300444C6173749100614368616E67653E0029D9265D00042200FF0552657472696576616C5175657374696F6E90D92429000BD0416E737765729080A46164616D
go
exec sp_executesql N'INSERT INTO [umbracoLog] ([userId],[NodeId],[entityType],[Datestamp],[logHeader],[logComment],[parameters]) VALUES (@0,@1,@2,@3,@4,@5,@6);SELECT SCOPE_IDENTITY();',N'@0 nvarchar(4000),@1 int,@2 nvarchar(4000),@3 datetime,@4 nvarchar(4000),@5 nvarchar(4000),@6 nvarchar(4000)',@0=NULL,@1=1149,@2=N'Member',@3='2021-07-13 17:03:50.920',@4=N'Save',@5=NULL,@6=NULL
go
```

**Edit** 

Whilst I though it was a DB object relationship issue it wasn't. Apologies, I didn't look carefully at the sequence I didn't apply the correct identity values when manually executing the statements.

When the application ran (and produced the above trace) my new member was not created. However, executing the above statements manually (taking care of the return id used in subsequent statements) I had the member created without issue.

### Steps to reproduce

DB migrated from v8 with existing members

Create view that automatically adds Windows authenticated use to Members list (snippet of code provided above)

### Expected result / actual result

New member should be created as expected

**Comment by adamhearn** [comment]
My existing DB has members with username's as per #10665

**Comment by adamhearn** [comment]
After investigating further it works if I swap out:

`_memberService.CreateMemberWithIdentity( User.Identity.Name, String.Format( "{0}@domain.com", empId ), System.Environment.UserName, "Member" );`

for

```
IMember member = _memberService.CreateMember( User.Identity.Name, String.Format( "{0}@domain.com", empId ), System.Environment.UserName, "Member" );
_memberService.Save( member, false );
```

or

```
MemberIdentityUser member = MemberIdentityUser.CreateNew( User.Identity.Name, "adam@domain.com", "Member", true, System.Environment.UserName );
await _memberManager.CreateAsync( member );
```

Could someone from the dev team please provide guidance on what is the "supported/preferred" method for creating a member programmatically.

**Comment by Shazwazza** [comment]
The same principal applies to both v8 and v9: 

* The IMemberService is for working with what is in the Database only. It has nothing to do with security implementations, that is up to .NET's security infrastructure. 
* For v8, the security infrastructure for members is the membership providers. For v9, the security infrastructure for members is ASP.NET Identity.

In v8, with the most up to date password security policies, you also cannot properly create a member with the IMemberService - because there is nothing there to assign or generate security policies like passwords, etc.  

The IMemberService will allow you to persist member data, but the password will just be set to the plain text you give the entity.

To create a new member with the correct security policies/passwords:

* In v9, use ASP.NET Identity using `IMemberManager`
* In v8, use the membership providers, or the MembershipHelper class

**Comment by Shazwazza** [comment]
That's not to say you haven't stumbled upon an issue though. `IMemberService.CreateMemberWithIdentity` should persist some data. Ideally there would be an integration test for the method `public IMember CreateMemberWithIdentity(string username, string email, string name, string memberTypeAlias)` to verify this. If you have time, a PR for such a thing in the `MemberServiceTests` would be wonderful :)

**Comment by adamhearn** [comment]
Thanks for the clarification of what API to use.

I've added the requested test. Not sure if the pull request will/can be accepted as the new test fails verifying the member is retrievable.

**Comment by Shazwazza** [comment]
Found the issue - this method was not completing it's scope/transaction so it was getting rolled back. Have fixed in rev: 91d58209827a6fb6f9809d866af4459f3e69827a
